Investigate TestSecondStepAfterFailure failing on CI sometimes

```
=== RUN   TestSecondStepAfterFailure

==================

WARNING: DATA RACE

Read at 0x00c0001203d8 by goroutine 55:

  github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep.TestSecondStepAfterFailure()

      /home/travis/gopath/src/github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep/reduction_test.go:61 +0x31a

  testing.tRunner()

      /home/travis/.gimme/versions/go1.13.12.linux.amd64/src/testing/testing.go:909 +0x199

Previous write at 0x00c0001203d8 by goroutine 91:

  github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep.(*Reducer).Halt()

      /home/travis/gopath/src/github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep/reducer.go:160 +0x92e

  github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep.(*Reducer).Halt-fm()

      /home/travis/gopath/src/github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep/reducer.go:148 +0x7d

  github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep.(*aggregator).collectVote()

      /home/travis/gopath/src/github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep/aggregator.go:76 +0x892

  github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep.(*Reducer).Collect()

      /home/travis/gopath/src/github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep/reducer.go:111 +0x8d1

  github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction.(*Helper).SendBatch.func1()

      /home/travis/gopath/src/github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/testutil.go:103 +0xa5

Goroutine 55 (running) created at:

  testing.(*T).Run()

      /home/travis/.gimme/versions/go1.13.12.linux.amd64/src/testing/testing.go:960 +0x651

  testing.runTests.func1()

      /home/travis/.gimme/versions/go1.13.12.linux.amd64/src/testing/testing.go:1202 +0xa6

  testing.tRunner()

      /home/travis/.gimme/versions/go1.13.12.linux.amd64/src/testing/testing.go:909 +0x199

  testing.runTests()

      /home/travis/.gimme/versions/go1.13.12.linux.amd64/src/testing/testing.go:1200 +0x521

  testing.(*M).Run()

      /home/travis/.gimme/versions/go1.13.12.linux.amd64/src/testing/testing.go:1117 +0x2ff

  main.main()

      _testmain.go:102 +0x337

Goroutine 91 (finished) created at:

  github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction.(*Helper).SendBatch()

      /home/travis/gopath/src/github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/testutil.go:102 +0x1b1

  github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep.TestSecondStepAfterFailure()

      /home/travis/gopath/src/github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep/reduction_test.go:50 +0x155

  testing.tRunner()

      /home/travis/.gimme/versions/go1.13.12.linux.amd64/src/testing/testing.go:909 +0x199

==================

--- FAIL: TestSecondStepAfterFailure (2.08s)

    reduction_test.go:61: 

        	Error Trace:	reduction_test.go:61

        	Error:      	Not equal: 

        	            	expected: 2000000000

        	            	actual  : 4000000000

        	Test:       	TestSecondStepAfterFailure

    testing.go:853: race detected during execution of test

FAIL

coverage: 88.2% of statements

FAIL	github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction/secondstep	3.987s
```

https://travis-ci.org/github/dusk-network/dusk-blockchain/jobs/699219538

--- a/pkg/core/consensus/reduction/firststep/aggregator.go
+++ b/pkg/core/consensus/reduction/firststep/aggregator.go
@@ -16,14 +16,14 @@ import (
 // The aggregator acts as a de facto storage unit for Reduction messages. Any message
 // it receives will be aggregated into a StepVotes struct, organized by block hash.
 // Once the key set for a StepVotes of a certain block hash reaches quorum, this
-// StepVotes is passed on to the Reducer by use of the `requestHalt` callback.
+// StepVotes is passed on to the Reducer by use of the `haltChan` channel.
 // An aggregator should be instantiated on a per-step basis and is no longer usable
-// after reaching quorum and calling `requestHalt`.
+// after reaching quorum and sending on `haltChan`.
 type aggregator struct {
-	requestHalt func([]byte, ...*message.StepVotes)
-	handler     *reduction.Handler
-	rpcBus      *rpcbus.RPCBus
-	finished    bool
+	haltChan chan<- reduction.HaltMsg
+	handler  *reduction.Handler
+	rpcBus   *rpcbus.RPCBus
+	finished bool
 
 	lock     sync.RWMutex
 	voteSets map[string]struct {
@@ -34,13 +34,13 @@ type aggregator struct {
 
 // newAggregator returns an instantiated aggregator, ready for use.
 func newAggregator(
-	requestHalt func([]byte, ...*message.StepVotes),
+	haltChan chan<- reduction.HaltMsg,
 	handler *reduction.Handler,
 	rpcBus *rpcbus.RPCBus) *aggregator {
 	return &aggregator{
-		requestHalt: requestHalt,
-		handler:     handler,
-		rpcBus:      rpcBus,
+		haltChan: haltChan,
+		handler:  handler,
+		rpcBus:   rpcBus,
 		voteSets: make(map[string]struct {
 			*message.StepVotes
 			sortedset.Cluster
@@ -102,12 +102,18 @@ func (a *aggregator) collectVote(ev message.Reduction) error {
 					WithField("round", hdr.Round).
 					WithField("step", hdr.Step).
 					Error("firststep_verifyCandidateBlock the candidate block failed")
-				a.requestHalt(emptyHash[:])
+				a.haltChan <- reduction.HaltMsg{
+					Hash: emptyHash[:],
+					Sv:   []*message.StepVotes{},
+				}
 				return nil
 			}
 		}
 
-		a.requestHalt(blockHash, sv.StepVotes)
+		a.haltChan <- reduction.HaltMsg{
+			Hash: blockHash,
+			Sv:   []*message.StepVotes{sv.StepVotes},
+		}
 	} else {
 		lg.
 			WithField("round", hdr.Round).

--- a/pkg/core/consensus/reduction/firststep/aggregator_test.go
+++ b/pkg/core/consensus/reduction/firststep/aggregator_test.go
@@ -4,7 +4,7 @@ import (
 	"testing"
 	"time"
 
-	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/message"
+	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction"
 	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/eventbus"
 	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/rpcbus"
 	"github.com/sirupsen/logrus"
@@ -14,16 +14,12 @@ import (
 // TestSuccessfulAggro tests that upon collection of a quorum of events, a valid StepVotes get produced
 func TestSuccessfulAggro(t *testing.T) {
 	eb, rbus := eventbus.New(), rpcbus.New()
-	hlp, hash := Kickstart(eb, rbus, 10, 1*time.Second)
+	hlp, hash := Kickstart(eb, rbus, 10, 1*time.Second, true)
 	evs := hlp.Spawn(hash)
 
-	res := make(chan error, 1)
-	test := func(hash []byte, svs ...*message.StepVotes) {
-		assert.Equal(t, hlp.Step(), svs[0].Step)
-		res <- hlp.Verify(hash, *svs[0], hlp.Step())
-	}
+	res := make(chan reduction.HaltMsg, 1)
 
-	aggregator := newAggregator(test, hlp.Handler, hlp.RBus)
+	aggregator := newAggregator(res, hlp.Handler, hlp.RBus)
 
 	for _, ev := range evs {
 		if !assert.NoError(t, aggregator.collectVote(ev)) {
@@ -31,26 +27,22 @@ func TestSuccessfulAggro(t *testing.T) {
 		}
 	}
 
-	err := <-res
-	assert.NoError(t, err)
+	m := <-res
+	assert.Equal(t, hlp.Step(), m.Sv[0].Step)
+	assert.NoError(t, hlp.Verify(m.Hash, *m.Sv[0], hlp.Step()))
 }
 
 // TestInvalidBlock tests that upon collection of a quorum of events, a valid StepVotes get produced
 func TestInvalidBlock(t *testing.T) {
 	logrus.SetLevel(logrus.FatalLevel)
 	eb, rbus := eventbus.New(), rpcbus.New()
-	hlp, hash := Kickstart(eb, rbus, 10, 1*time.Second)
+	hlp, hash := Kickstart(eb, rbus, 10, 1*time.Second, true)
 	hlp.FailOnVerification(true)
 	evs := hlp.Spawn(hash)
 
-	res := make(chan struct{}, 1)
-	test := func(hash []byte, svs ...*message.StepVotes) {
-		assert.Equal(t, emptyHash[:], hash)
-		assert.Equal(t, 0, len(svs))
-		res <- struct{}{}
-	}
+	res := make(chan reduction.HaltMsg, 1)
 
-	aggregator := newAggregator(test, hlp.Handler, hlp.RBus)
+	aggregator := newAggregator(res, hlp.Handler, hlp.RBus)
 
 	for _, ev := range evs {
 		if !assert.NoError(t, aggregator.collectVote(ev)) {
@@ -58,25 +50,22 @@ func TestInvalidBlock(t *testing.T) {
 		}
 	}
 
-	<-res
+	m := <-res
+	assert.Equal(t, emptyHash[:], m.Hash)
+	assert.Equal(t, 0, len(m.Sv))
 }
 
 // Test that a valid stepvotes is produced when a candidate block for
 // a given hash is not found.
 func TestCandidateNotFound(t *testing.T) {
 	eb, rbus := eventbus.New(), rpcbus.New()
-	hlp, hash := Kickstart(eb, rbus, 10, 1*time.Second)
+	hlp, hash := Kickstart(eb, rbus, 10, 1*time.Second, true)
 	hlp.FailOnFetching(true)
 	evs := hlp.Spawn(hash)
 
-	res := make(chan struct{}, 1)
-	test := func(hash []byte, svs ...*message.StepVotes) {
-		assert.Equal(t, emptyHash[:], hash)
-		assert.Equal(t, 0, len(svs))
-		res <- struct{}{}
-	}
+	res := make(chan reduction.HaltMsg, 1)
 
-	aggregator := newAggregator(test, hlp.Handler, hlp.RBus)
+	aggregator := newAggregator(res, hlp.Handler, hlp.RBus)
 
 	for _, ev := range evs {
 		if !assert.NoError(t, aggregator.collectVote(ev)) {
@@ -84,5 +73,7 @@ func TestCandidateNotFound(t *testing.T) {
 		}
 	}
 
-	<-res
+	m := <-res
+	assert.Equal(t, emptyHash[:], m.Hash)
+	assert.Equal(t, 0, len(m.Sv))
 }

--- a/pkg/core/consensus/reduction/firststep/aggregator_test.go
+++ b/pkg/core/consensus/reduction/firststep/aggregator_test.go
@@ -14,7 +14,7 @@ import (
 // TestSuccessfulAggro tests that upon collection of a quorum of events, a valid StepVotes get produced
 func TestSuccessfulAggro(t *testing.T) {
 	eb, rbus := eventbus.New(), rpcbus.New()
-	hlp, hash := Kickstart(eb, rbus, 10, 1*time.Second, true)
+	hlp, hash := KickstartConcurrent(eb, rbus, 10, 1*time.Second)
 	evs := hlp.Spawn(hash)
 
 	res := make(chan reduction.HaltMsg, 1)
@@ -36,7 +36,7 @@ func TestSuccessfulAggro(t *testing.T) {
 func TestInvalidBlock(t *testing.T) {
 	logrus.SetLevel(logrus.FatalLevel)
 	eb, rbus := eventbus.New(), rpcbus.New()
-	hlp, hash := Kickstart(eb, rbus, 10, 1*time.Second, true)
+	hlp, hash := KickstartConcurrent(eb, rbus, 10, 1*time.Second)
 	hlp.FailOnVerification(true)
 	evs := hlp.Spawn(hash)
 
@@ -59,7 +59,7 @@ func TestInvalidBlock(t *testing.T) {
 // a given hash is not found.
 func TestCandidateNotFound(t *testing.T) {
 	eb, rbus := eventbus.New(), rpcbus.New()
-	hlp, hash := Kickstart(eb, rbus, 10, 1*time.Second, true)
+	hlp, hash := KickstartConcurrent(eb, rbus, 10, 1*time.Second)
 	hlp.FailOnFetching(true)
 	evs := hlp.Spawn(hash)
 

--- a/pkg/core/consensus/reduction/firststep/reducer.go
+++ b/pkg/core/consensus/reduction/firststep/reducer.go
@@ -33,15 +33,20 @@ type Reducer struct {
 	timeOut    time.Duration
 	Timer      *reduction.Timer
 	round      uint64
+
+	haltChan chan reduction.HaltMsg
+	quitChan chan struct{}
 }
 
 // NewComponent returns an uninitialized reduction component.
 func NewComponent(broker eventbus.Broker, rpcBus *rpcbus.RPCBus, keys key.Keys, timeOut time.Duration) reduction.Reducer {
 	return &Reducer{
-		broker:  broker,
-		rpcBus:  rpcBus,
-		keys:    keys,
-		timeOut: timeOut,
+		broker:   broker,
+		rpcBus:   rpcBus,
+		keys:     keys,
+		timeOut:  timeOut,
+		haltChan: make(chan reduction.HaltMsg, 2),
+		quitChan: make(chan struct{}, 1),
 	}
 }
 
@@ -52,7 +57,7 @@ func (r *Reducer) Initialize(eventPlayer consensus.EventPlayer, signer consensus
 	r.eventPlayer = eventPlayer
 	r.signer = signer
 	r.handler = reduction.NewHandler(r.keys, ru.P)
-	r.Timer = reduction.NewTimer(r.Halt)
+	r.Timer = reduction.NewTimer(r.haltChan)
 	r.round = ru.Round
 
 	bestScoreSubscriber := consensus.TopicListener{
@@ -82,6 +87,7 @@ func (r *Reducer) ID() uint32 {
 func (r *Reducer) Finalize() {
 	r.eventPlayer.Pause(r.reductionID)
 	r.Timer.Stop()
+	r.quitChan <- struct{}{}
 }
 
 // Collect forwards Reduction to the aggregator
@@ -115,8 +121,15 @@ func (r *Reducer) startReduction() {
 		"id":      r.reductionID,
 		"timeout": r.timeOut / time.Second,
 	}).Debugln("firststep, startReduction")
+
+	// Clear out haltChan
+	for len(r.haltChan) > 0 {
+		<-r.haltChan
+	}
+
+	go r.collectHalt()
 	r.Timer.Start(r.timeOut)
-	r.aggregator = newAggregator(r.Halt, r.handler, r.rpcBus)
+	r.aggregator = newAggregator(r.haltChan, r.handler, r.rpcBus)
 }
 
 func (r *Reducer) sendReduction(step uint8, hash []byte) {
@@ -132,6 +145,16 @@ func (r *Reducer) sendReduction(step uint8, hash []byte) {
 
 	if err := r.signer.Gossip(msg, r.ID()); err != nil {
 		lg.WithField("category", "BUG").WithError(err).Errorln("error in sending authenticated Reduction")
+	}
+}
+
+// collectHalt is called by the Aggregator and the Timer through a channel, and ensures that only
+// one `Halt` call is made per step.
+func (r *Reducer) collectHalt() {
+	select {
+	case m := <-r.haltChan:
+		r.Halt(m.Hash, m.Sv)
+	case <-r.quitChan:
 	}
 }
 
@@ -156,7 +179,7 @@ func (s StepVotesMsgFactory) Create(sender []byte, round uint64, step uint8) con
 
 // Halt will end the first step of reduction, and forwards whatever result it received
 // on the StepVotes topic.
-func (r *Reducer) Halt(hash []byte, svs ...*message.StepVotes) {
+func (r *Reducer) Halt(hash []byte, svs []*message.StepVotes) {
 	var svm message.StepVotesMsg
 	lg.
 		WithField("id", r.reductionID).

--- a/pkg/core/consensus/reduction/firststep/reduction_test.go
+++ b/pkg/core/consensus/reduction/firststep/reduction_test.go
@@ -2,6 +2,7 @@ package firststep
 
 import (
 	"bytes"
+	"runtime"
 	"testing"
 	"time"
 
@@ -65,7 +66,7 @@ func TestMoreSteps(t *testing.T) {
 func TestFirstStepTimeOut(t *testing.T) {
 	bus, rpcBus := eventbus.New(), rpcbus.New()
 	timeOut := 100 * time.Millisecond
-	hlp, _ := Kickstart(bus, rpcBus, 50, timeOut)
+	hlp, _ := Kickstart(bus, rpcBus, 50, timeOut, true)
 
 	// Wait for resulting StepVotes
 	svMsg := <-hlp.StepVotesChan
@@ -85,7 +86,7 @@ func TestFirstStepTimeOut(t *testing.T) {
 
 func BenchmarkFirstStep(b *testing.B) {
 	bus, rpcBus := eventbus.New(), rpcbus.New()
-	hlp, hash := Kickstart(bus, rpcBus, 50, 1*time.Second)
+	hlp, hash := Kickstart(bus, rpcBus, 50, 1*time.Second, true)
 	b.ResetTimer()
 	b.StopTimer()
 	for i := 0; i < b.N; i++ {
@@ -99,4 +100,22 @@ func BenchmarkFirstStep(b *testing.B) {
 		hash, _ = crypto.RandEntropy(32)
 		hlp.ActivateReduction(hash)
 	}
+}
+
+// Test that we properly clean up after calling Finalize.
+// TODO: trap eventual errors
+func TestFinalize(t *testing.T) {
+	numGRBefore := runtime.NumGoroutine()
+	// Create a set of 100 agreement components, and finalize them immediately
+	for i := 0; i < 100; i++ {
+		bus, rpcBus := eventbus.New(), rpcbus.New()
+		hlp, _ := Kickstart(bus, rpcBus, 50, 1*time.Second, false)
+
+		hlp.Reducer.Finalize()
+	}
+
+	// Ensure we have freed up all of the resources associated with these components
+	numGRAfter := runtime.NumGoroutine()
+	// We should have roughly the same amount of goroutines
+	assert.InDelta(t, numGRBefore, numGRAfter, 10.0)
 }

--- a/pkg/core/consensus/reduction/firststep/reduction_test.go
+++ b/pkg/core/consensus/reduction/firststep/reduction_test.go
@@ -66,7 +66,7 @@ func TestMoreSteps(t *testing.T) {
 func TestFirstStepTimeOut(t *testing.T) {
 	bus, rpcBus := eventbus.New(), rpcbus.New()
 	timeOut := 100 * time.Millisecond
-	hlp, _ := Kickstart(bus, rpcBus, 50, timeOut, true)
+	hlp, _ := KickstartConcurrent(bus, rpcBus, 50, timeOut)
 
 	// Wait for resulting StepVotes
 	svMsg := <-hlp.StepVotesChan
@@ -86,7 +86,7 @@ func TestFirstStepTimeOut(t *testing.T) {
 
 func BenchmarkFirstStep(b *testing.B) {
 	bus, rpcBus := eventbus.New(), rpcbus.New()
-	hlp, hash := Kickstart(bus, rpcBus, 50, 1*time.Second, true)
+	hlp, hash := KickstartConcurrent(bus, rpcBus, 50, 1*time.Second)
 	b.ResetTimer()
 	b.StopTimer()
 	for i := 0; i < b.N; i++ {
@@ -109,7 +109,7 @@ func TestFinalize(t *testing.T) {
 	// Create a set of 100 agreement components, and finalize them immediately
 	for i := 0; i < 100; i++ {
 		bus, rpcBus := eventbus.New(), rpcbus.New()
-		hlp, _ := Kickstart(bus, rpcBus, 50, 1*time.Second, false)
+		hlp, _ := Kickstart(bus, rpcBus, 50, 1*time.Second)
 
 		hlp.Reducer.Finalize()
 	}

--- a/pkg/core/consensus/reduction/firststep/testutil.go
+++ b/pkg/core/consensus/reduction/firststep/testutil.go
@@ -26,7 +26,7 @@ type Helper struct {
 }
 
 // NewHelper creates a Helper
-func NewHelper(eb *eventbus.EventBus, rpcbus *rpcbus.RPCBus, provisioners int, timeOut time.Duration) *Helper {
+func NewHelper(eb *eventbus.EventBus, rpcbus *rpcbus.RPCBus, provisioners int, timeOut time.Duration, startGoroutines bool) *Helper {
 	hlp := &Helper{
 		Helper:             reduction.NewHelper(eb, rpcbus, provisioners, CreateReducer, timeOut),
 		StepVotesChan:      make(chan message.Message, 1),
@@ -34,8 +34,10 @@ func NewHelper(eb *eventbus.EventBus, rpcbus *rpcbus.RPCBus, provisioners int, t
 		failOnVerification: false,
 	}
 
-	go hlp.provideCandidateBlock()
-	go hlp.processCandidateVerificationRequest()
+	if startGoroutines {
+		go hlp.provideCandidateBlock()
+		go hlp.processCandidateVerificationRequest()
+	}
 	hlp.createResultChan()
 	return hlp
 }
@@ -120,8 +122,8 @@ func (hlp *Helper) NextBatch() []byte {
 }
 
 // Kickstart a Helper without sending any reduction event
-func Kickstart(eb *eventbus.EventBus, rpcbus *rpcbus.RPCBus, nr int, timeOut time.Duration) (*Helper, []byte) {
-	h := NewHelper(eb, rpcbus, nr, timeOut)
+func Kickstart(eb *eventbus.EventBus, rpcbus *rpcbus.RPCBus, nr int, timeOut time.Duration, startGoroutines bool) (*Helper, []byte) {
+	h := NewHelper(eb, rpcbus, nr, timeOut, startGoroutines)
 	roundUpdate := consensus.MockRoundUpdate(h.Round, h.P)
 	h.Initialize(roundUpdate)
 	hash, _ := crypto.RandEntropy(32)
@@ -131,7 +133,7 @@ func Kickstart(eb *eventbus.EventBus, rpcbus *rpcbus.RPCBus, nr int, timeOut tim
 
 // ProduceFirstStepVotes encapsulates the process of creating and forwarding Reduction events
 func ProduceFirstStepVotes(eb *eventbus.EventBus, rpcbus *rpcbus.RPCBus, nr int, timeOut time.Duration) (*Helper, []byte) {
-	hlp, hash := Kickstart(eb, rpcbus, nr, timeOut)
+	hlp, hash := Kickstart(eb, rpcbus, nr, timeOut, true)
 	hlp.SendBatch(hash)
 	return hlp, hash
 }

--- a/pkg/core/consensus/reduction/firststep/testutil.go
+++ b/pkg/core/consensus/reduction/firststep/testutil.go
@@ -123,19 +123,33 @@ func (hlp *Helper) NextBatch() []byte {
 	return blockHash
 }
 
-// Kickstart a Helper without sending any reduction event
-func Kickstart(eb *eventbus.EventBus, rpcbus *rpcbus.RPCBus, nr int, timeOut time.Duration, startGoroutines bool) (*Helper, []byte) {
-	h := NewHelper(eb, rpcbus, nr, timeOut, startGoroutines)
+// Kickstart a Helper without sending any reduction event, and without starting goroutines to
+// intercept RPCBus calls.
+func Kickstart(eb *eventbus.EventBus, rpcbus *rpcbus.RPCBus, nr int, timeOut time.Duration) (*Helper, []byte) {
+	h := NewHelper(eb, rpcbus, nr, timeOut, false)
+	hash := kickstart(h)
+	return h, hash
+}
+
+// KickstartConcurrent kickstarts a Helper without sending any reduction event, starting
+// goroutines to intercept RPCBus calls.
+func KickstartConcurrent(eb *eventbus.EventBus, rpcbus *rpcbus.RPCBus, nr int, timeOut time.Duration) (*Helper, []byte) {
+	h := NewHelper(eb, rpcbus, nr, timeOut, true)
+	hash := kickstart(h)
+	return h, hash
+}
+
+func kickstart(h *Helper) []byte {
 	roundUpdate := consensus.MockRoundUpdate(h.Round, h.P)
 	h.Initialize(roundUpdate)
 	hash, _ := crypto.RandEntropy(32)
 	h.ActivateReduction(hash)
-	return h, hash
+	return hash
 }
 
 // ProduceFirstStepVotes encapsulates the process of creating and forwarding Reduction events
 func ProduceFirstStepVotes(eb *eventbus.EventBus, rpcbus *rpcbus.RPCBus, nr int, timeOut time.Duration) (*Helper, []byte) {
-	hlp, hash := Kickstart(eb, rpcbus, nr, timeOut, true)
+	hlp, hash := KickstartConcurrent(eb, rpcbus, nr, timeOut)
 	hlp.SendBatch(hash)
 	return hlp, hash
 }

--- a/pkg/core/consensus/reduction/firststep/testutil.go
+++ b/pkg/core/consensus/reduction/firststep/testutil.go
@@ -25,7 +25,9 @@ type Helper struct {
 	failOnVerification bool
 }
 
-// NewHelper creates a Helper
+// NewHelper creates a Helper used for testing the first step Reducer.
+// `startGoroutines` can be specified to simultaneously launch goroutines
+// that intercept RPC calls made by the first step Reducer.
 func NewHelper(eb *eventbus.EventBus, rpcbus *rpcbus.RPCBus, provisioners int, timeOut time.Duration, startGoroutines bool) *Helper {
 	hlp := &Helper{
 		Helper:             reduction.NewHelper(eb, rpcbus, provisioners, CreateReducer, timeOut),

--- a/pkg/core/consensus/reduction/reducer.go
+++ b/pkg/core/consensus/reduction/reducer.go
@@ -1,6 +1,15 @@
 package reduction
 
-import "github.com/dusk-network/dusk-blockchain/pkg/core/consensus"
+import (
+	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/message"
+)
+
+// HaltMsg is used to communicate between reducers, and aggregators and timers.
+type HaltMsg struct {
+	Hash []byte
+	Sv   []*message.StepVotes
+}
 
 // Reducer abstracts the first and second step Reduction components
 type Reducer interface {

--- a/pkg/core/consensus/reduction/reduction_test.go
+++ b/pkg/core/consensus/reduction/reduction_test.go
@@ -61,10 +61,15 @@ func TestCorrectHeader(t *testing.T) {
 		t.FailNow()
 	}
 
-	// We discard the message in the middle, as it will be an Agreement message,
-	// as a result of the emptying of the queue resulting on quorum.
-	<-gossipChan
-	r2 := <-gossipChan
+	// There is no guaranteed order between the agreement message, and the second reduction
+	// message. So, we attempt to get the right one before moving on.
+	var r2 message.Message
+	for {
+		r2 = <-gossipChan
+		if r2.Category() == topics.Reduction {
+			break
+		}
+	}
 
 	// Retrieve headers from both reduction messages
 	p2 := r2.Payload().(message.SafeBuffer)
@@ -111,7 +116,7 @@ func collectEvents(t *testing.T, c *consensus.Coordinator, evs []message.Reducti
 }
 
 func wireReduction(t *testing.T, bus *eventbus.EventBus, rpcBus *rpcbus.RPCBus) (*consensus.Coordinator, *firststep.Helper) {
-	hlp := firststep.NewHelper(bus, rpcBus, 10, 1*time.Second)
+	hlp := firststep.NewHelper(bus, rpcBus, 10, 1*time.Second, true)
 	f1 := firststep.NewFactory(bus, rpcBus, hlp.Keys[0], 1*time.Second)
 	f2 := secondstep.NewFactory(bus, rpcBus, hlp.Keys[0], 1*time.Second)
 	c := consensus.Start(bus, hlp.Keys[0], f1, f2)

--- a/pkg/core/consensus/reduction/secondstep/aggregator.go
+++ b/pkg/core/consensus/reduction/secondstep/aggregator.go
@@ -11,11 +11,11 @@ import (
 // The aggregator acts as a de facto storage unit for Reduction messages. Any message
 // it receives will be aggregated into a StepVotes struct, organized by block hash.
 // Once the key set for a StepVotes of a certain block hash reaches quorum, this
-// StepVotes is passed on to the Reducer by use of the `requestHalt` callback.
+// StepVotes is passed on to the Reducer by use of the `haltChan` channel.
 // An aggregator should be instantiated on a per-step basis and is no longer usable
-// after reaching quorum and calling `requestHalt`.
+// after reaching quorum and sending on `haltChan`.
 type aggregator struct {
-	requestHalt    func([]byte, ...*message.StepVotes)
+	haltChan       chan<- reduction.HaltMsg
 	handler        *reduction.Handler
 	firstStepVotes *message.StepVotes
 	finished       bool
@@ -29,12 +29,12 @@ type aggregator struct {
 
 // newAggregator returns an instantiated aggregator, ready for use.
 func newAggregator(
-	requestHalt func([]byte, ...*message.StepVotes),
+	haltChan chan<- reduction.HaltMsg,
 	handler *reduction.Handler,
 	firstStepVotes *message.StepVotes) *aggregator {
 
 	return &aggregator{
-		requestHalt:    requestHalt,
+		haltChan:       haltChan,
 		handler:        handler,
 		firstStepVotes: firstStepVotes,
 		voteSets: make(map[string]struct {
@@ -84,7 +84,10 @@ func (a *aggregator) collectVote(ev message.Reduction) error {
 			Debug("secondstep_quorum_reached")
 		a.finished = true
 		a.addBitSet(sv.StepVotes, sv.Cluster, hdr.Round, hdr.Step)
-		a.requestHalt(hdr.BlockHash, a.firstStepVotes, sv.StepVotes)
+		a.haltChan <- reduction.HaltMsg{
+			Hash: hdr.BlockHash,
+			Sv:   []*message.StepVotes{a.firstStepVotes, sv.StepVotes},
+		}
 	} else {
 		lg.
 			WithField("round", hdr.Round).

--- a/pkg/core/consensus/reduction/secondstep/aggregator_test.go
+++ b/pkg/core/consensus/reduction/secondstep/aggregator_test.go
@@ -4,6 +4,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/message"
 	"github.com/stretchr/testify/assert"
 )
@@ -12,13 +13,10 @@ func TestSuccessfulAggro(t *testing.T) {
 	hlp, hash := Kickstart(10, 1*time.Second)
 	evs := hlp.Spawn(hash)
 
-	res := make(chan error, 1)
-	test := func(hash []byte, svs ...*message.StepVotes) {
-		res <- hlp.Verify(hash, *svs[1], hlp.Step())
-	}
+	res := make(chan reduction.HaltMsg, 1)
 
 	var sv *message.StepVotes
-	aggregator := newAggregator(test, hlp.Handler, sv)
+	aggregator := newAggregator(res, hlp.Handler, sv)
 
 	go func() {
 		for _, r := range evs {
@@ -28,6 +26,6 @@ func TestSuccessfulAggro(t *testing.T) {
 		}
 	}()
 
-	err := <-res
-	assert.NoError(t, err)
+	m := <-res
+	assert.NoError(t, hlp.Verify(m.Hash, *m.Sv[1], hlp.Step()))
 }

--- a/pkg/core/consensus/reduction/secondstep/reducer.go
+++ b/pkg/core/consensus/reduction/secondstep/reducer.go
@@ -39,15 +39,20 @@ type Reducer struct {
 	timeOut    time.Duration
 	timer      *reduction.Timer
 	round      uint64
+
+	haltChan chan reduction.HaltMsg
+	quitChan chan struct{}
 }
 
 // NewComponent returns an uninitialized reduction component.
 func NewComponent(broker eventbus.Broker, rpcBus *rpcbus.RPCBus, keys key.Keys, timeOut time.Duration) reduction.Reducer {
 	return &Reducer{
-		broker:  broker,
-		rpcBus:  rpcBus,
-		keys:    keys,
-		timeOut: timeOut,
+		broker:   broker,
+		rpcBus:   rpcBus,
+		keys:     keys,
+		timeOut:  timeOut,
+		haltChan: make(chan reduction.HaltMsg, 2),
+		quitChan: make(chan struct{}, 1),
 	}
 }
 
@@ -58,7 +63,7 @@ func (r *Reducer) Initialize(eventPlayer consensus.EventPlayer, signer consensus
 	r.eventPlayer = eventPlayer
 	r.signer = signer
 	r.handler = reduction.NewHandler(r.keys, ru.P)
-	r.timer = reduction.NewTimer(r.Halt)
+	r.timer = reduction.NewTimer(r.haltChan)
 	r.round = ru.Round
 
 	stepVotesSubscriber := consensus.TopicListener{
@@ -88,6 +93,7 @@ func (r *Reducer) ID() uint32 {
 func (r *Reducer) Finalize() {
 	r.eventPlayer.Pause(r.reductionID)
 	r.timer.Stop()
+	r.quitChan <- struct{}{}
 }
 
 // Collect complies with the consensus.Component interface. It gathers
@@ -123,8 +129,15 @@ func (r *Reducer) startReduction(sv message.StepVotesMsg) {
 		"id":      r.reductionID,
 		"timeout": r.timeOut / time.Second,
 	}).Debugln("startReduction")
+
+	// Clear out haltChan
+	for len(r.haltChan) > 0 {
+		<-r.haltChan
+	}
+
+	go r.collectHalt()
 	r.timer.Start(r.timeOut)
-	r.aggregator = newAggregator(r.Halt, r.handler, &sv.StepVotes)
+	r.aggregator = newAggregator(r.haltChan, r.handler, &sv.StepVotes)
 }
 
 // sendReduction creates a new Reduction message and Gossip it externally
@@ -148,9 +161,19 @@ func (r *Reducer) sendReduction(step uint8, hash []byte) error {
 
 var restartFactory = consensus.Restarter{}
 
+// collectHalt is called by the Aggregator and the Timer through a channel, and ensures that only
+// one `Halt` call is made per step.
+func (r *Reducer) collectHalt() {
+	select {
+	case m := <-r.haltChan:
+		r.Halt(m.Hash, m.Sv)
+	case <-r.quitChan:
+	}
+}
+
 // Halt is used by either the Aggregator in case of successful reduction or the timer in case of a timeout.
 // In the latter case no agreement message is pushed forward
-func (r *Reducer) Halt(hash []byte, b ...*message.StepVotes) {
+func (r *Reducer) Halt(hash []byte, b []*message.StepVotes) {
 	lg.
 		WithField("len_step", len(b)).
 		WithField("id", r.reductionID).

--- a/pkg/core/consensus/reduction/secondstep/reduction_test.go
+++ b/pkg/core/consensus/reduction/secondstep/reduction_test.go
@@ -1,6 +1,7 @@
 package secondstep
 
 import (
+	"runtime"
 	"testing"
 	"time"
 
@@ -38,7 +39,7 @@ func TestSecondStep(t *testing.T) {
 }
 
 func TestSecondStepAfterFailure(t *testing.T) {
-	timeOut := 1000 * time.Millisecond
+	timeOut := 100 * time.Millisecond
 	hlp, hash := Kickstart(50, timeOut)
 
 	// Start the first step
@@ -61,4 +62,26 @@ func TestSecondStepAfterFailure(t *testing.T) {
 		assert.Equal(t, timeOut*2, hlp.Reducer.(*Reducer).timeOut)
 		// Success
 	}
+}
+
+// Test that we properly clean up after calling Finalize.
+// TODO: trap eventual errors
+func TestFinalize(t *testing.T) {
+	numGRBefore := runtime.NumGoroutine()
+	// Create a set of 100 reduction components, and finalize them immediately
+	for i := 0; i < 100; i++ {
+		hlp, hash := Kickstart(50, 1*time.Second)
+
+		// Start the first step
+		if err := hlp.ActivateReduction(hash, message.StepVotes{}); err != nil {
+			t.Fatal(err)
+		}
+
+		hlp.Reducer.Finalize()
+	}
+
+	// Ensure we have freed up all of the resources associated with these components
+	numGRAfter := runtime.NumGoroutine()
+	// We should have roughly the same amount of goroutines
+	assert.InDelta(t, numGRBefore, numGRAfter, 10.0)
 }

--- a/pkg/core/consensus/reduction/timer.go
+++ b/pkg/core/consensus/reduction/timer.go
@@ -13,15 +13,15 @@ var emptyHash [32]byte
 // Timer sets a timeout for collecting reduction messages if no quorum is
 // reached after a while
 type Timer struct {
-	requestHalt func([]byte, ...*message.StepVotes)
-	lock        sync.RWMutex
-	t           *time.Timer
+	haltChan chan<- HaltMsg
+	lock     sync.RWMutex
+	t        *time.Timer
 }
 
 // NewTimer instantiates a new Timer
-func NewTimer(requestHalt func([]byte, ...*message.StepVotes)) *Timer {
+func NewTimer(haltChan chan<- HaltMsg) *Timer {
 	return &Timer{
-		requestHalt: requestHalt,
+		haltChan: haltChan,
 	}
 }
 
@@ -44,5 +44,8 @@ func (t *Timer) Stop() {
 // Trigger the timeout and requests a halt with empty block hash
 func (t *Timer) Trigger() {
 	log.WithField("process", "reduction timer").Debugln("timer triggered")
-	t.requestHalt(emptyHash[:])
+	t.haltChan <- HaltMsg{
+		Hash: emptyHash[:],
+		Sv:   []*message.StepVotes{},
+	}
 }

--- a/pkg/core/consensus/reduction/timer_test.go
+++ b/pkg/core/consensus/reduction/timer_test.go
@@ -4,13 +4,13 @@ import (
 	"testing"
 
 	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction"
-	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/message"
 	"github.com/stretchr/testify/assert"
 )
 
 // Ensure that stopping a timer which was never started does not result
 // in a panic.
 func TestStopNilTimer(t *testing.T) {
-	timer := reduction.NewTimer(func([]byte, ...*message.StepVotes) {})
+	c := make(chan reduction.HaltMsg, 1)
+	timer := reduction.NewTimer(c)
 	assert.NotPanics(t, timer.Stop)
 }

